lib/ClangImporter/ImportType.cpp

```diff
@@ -2227,6 +2227,157 @@ ImportedType ClangImporter::Implementation::importFunctionParamsAndReturnType(
+    Optional<unsigned> completionHandlerErrorParamIndex,
+    ArrayRef<GenericTypeParamDecl *> genericParams,
+    llvm::function_ref<void(Diagnostic &&)> addImportDiagnosticFn,
+    bool &isInOut, bool &isParamTypeImplicitlyUnwrapped) {
+  auto paramTy = param->getType();
+
+  // Import the parameter type into Swift.
+  Type swiftParamTy;
+
+  // Sometimes we import unavailable typedefs as enums. If that's the case,
+  // use the enum, not the typedef here.
+  if (auto typedefType = dyn_cast<clang::TypedefType>(paramTy.getTypePtr())) {
+    if (isUnavailableInSwift(typedefType->getDecl())) {
+      if (auto clangEnum =
+              findAnonymousEnumForTypedef(SwiftContext, typedefType)) {
+        // If this fails, it means that we need a stronger predicate for
+        // determining the relationship between an enum and typedef.
+        assert(clangEnum.getValue()
+                   ->getIntegerType()
+                   ->getCanonicalTypeInternal() ==
+               typedefType->getCanonicalTypeInternal());
+        if (auto swiftEnum = importDecl(*clangEnum, CurrentVersion)) {
+          swiftParamTy = cast<NominalTypeDecl>(swiftEnum)->getDeclaredType();
+        }
+      }
+    }
+  } else if (isa<clang::PointerType>(paramTy) &&
+             isa<clang::TemplateTypeParmType>(paramTy->getPointeeType())) {
+    auto pointeeType = paramTy->getPointeeType();
+    auto templateParamType = cast<clang::TemplateTypeParmType>(pointeeType);
+    PointerTypeKind pointerKind = pointeeType.getQualifiers().hasConst()
+                                      ? PTK_UnsafePointer
+                                      : PTK_UnsafeMutablePointer;
+    auto attrs = getImportTypeAttrs(param, /*isParam=*/true);
+    auto genericType = findGenericTypeInGenericDecls(
+        *this, templateParamType, genericParams, attrs, addImportDiagnosticFn);
+    swiftParamTy = genericType->wrapInPointer(pointerKind);
+    if (!swiftParamTy)
+      return None;
+  } else if (isa<clang::ReferenceType>(paramTy) &&
+             isa<clang::TemplateTypeParmType>(paramTy->getPointeeType())) {
+    auto templateParamType =
+        cast<clang::TemplateTypeParmType>(paramTy->getPointeeType());
+    auto attrs = getImportTypeAttrs(param, /*isParam=*/true);
+    swiftParamTy = findGenericTypeInGenericDecls(
+        *this, templateParamType, genericParams, attrs, addImportDiagnosticFn);
+    if (!paramTy->getPointeeType().isConstQualified())
+      isInOut = true;
+  } else if (auto *templateParamType =
+                 dyn_cast<clang::TemplateTypeParmType>(paramTy)) {
+    auto attrs = getImportTypeAttrs(param, /*isParam=*/true);
+    swiftParamTy = findGenericTypeInGenericDecls(
+        *this, templateParamType, genericParams, attrs, addImportDiagnosticFn);
+  } else if (auto refType = dyn_cast<clang::ReferenceType>(paramTy)) {
+    // We don't support reference type to a dependent type, just bail.
+    if (refType->getPointeeType()->isDependentType()) {
+      return None;
```

Comment:
Why are we missing

```
       addImportDiagnostic(
            param, Diagnostic(diag::parameter_type_not_imported, param),
            param->getSourceRange().getBegin());
```

here?

Reply:
All the `addImportDiagnostic` calls should have been moved outside `importParameterType`. The equivalent is in Line 2418 of the new version. This means that other `return None` of the new function will add the right diagnostics (like in Line 2284, which in the previous version will not have included diagnostics).